add --cluster-driver and --cluster-opt

vieux opened the pull request:
![http://cl.ly/image/0y1C0O0t3q1M/Screen%20Shot%202015-05-12%20at%2017.50.28.png](http://cl.ly/image/0y1C0O0t3q1M/Screen%20Shot%202015-05-12%20at%2017.50.28.png)

See #633

vieux commented:
I opted for less changes as possible, we could refactor the code later

vieux commented:
ping @aluzzardi

vieux commented:
ping @jimenez @tnachen does this looks good to you ?

tnachen commented:
@vieux I see, wrapping everything in cluster-opts is fine, I think this change makes sense, +1

vieux commented:
I pushed another commit with a couple of changes:

* DriverOpts is now a type with a few getters as suggested by @aluzzardi 
* I figured that the heartbeat is also swarm specific
* I removed the option struct because I hated it since I introduced it
 * TLSConfig is now a parameter
 * discovery is now injected on the fly (it's good because for mesos it will be like `mesos.master`)

@aluzzardi let me know what you think

jimenez commented:
LGTM

abronan commented:
LGTM

vieux commented:
@aluzzardi I added some test, the CI is green now :)

aluzzardi commented on cli/manage.go:
```diff
@@ -120,19 +119,11 @@ func manage(c *cli.Context) {
-		OvercommitRatio: c.Float64("overcommit"),
-		Discovery:       dflag,
-		Heartbeat:       hb,
+	cluster, err := swarm.NewCluster(sched, store, tlsConfig, append(c.StringSlice("cluster-opt"), "swarm.discovery="+dflag))
```
Is discovery `swarm` only? It's kind of weird to pass it like that

vieux replied:
yes, we could add another argument to NewCluster to pass it if you think it's better.

aluzzardi commented on cluster/options.go:
```diff
@@ -1,11 +1,47 @@
+type DriverOpts []string
+
+// String returns a string from the driver options
+func (do DriverOpts) String(key string) (bool, string) {
```
nit: To be consistent with `map`, `bool` should be the last argument:

```go
str, ok := opts.String("foo")
```

aluzzardi commented on cluster/options.go:
```diff
@@ -1,11 +1,47 @@
+}
+
+// Int returns an int64 from the driver options
+func (do DriverOpts) Int(key string) (bool, int64) {
```
Same

aluzzardi commented on cluster/options.go:
```diff
@@ -1,11 +1,47 @@
+}
+
+// Uint returns an int64 from the driver options
+func (do DriverOpts) Uint(key string) (bool, uint64) {
```
Same

aluzzardi commented on cluster/options.go:
```diff
@@ -1,11 +1,47 @@
+}
+
+// Float returns a float64 from the driver options
+func (do DriverOpts) Float(key string) (bool, float64) {
```
Same

aluzzardi commented on cluster/options_test.go:
```diff
@@ -0,0 +1,97 @@
+	assert.True(t, ok)
+	assert.Equal(t, val, "0.6")
+
+	ok, val = opts.String("foo5")
```
tiny little nit, feel free to ignore: Instead of `foo5` I'd use `invalid` or `non-existent`. It's not obvious that `foo5` doesn't exist

aluzzardi commented on cluster/swarm/cluster.go:
```diff
@@ -26,24 +27,37 @@ type Cluster struct {
 	// get the list of entries from the discovery service
 	go func() {
-		d, err := discovery.New(options.Discovery, options.Heartbeat)
+		d, err := discovery.New(cluster.discovery, cluster.heartbeat)
```
Entirely unrelated to this PR: `heartbeat` should be a discovery flag, not a cluster flag.

It only makes sense for `token` (and `file` but it should be called something like `refresh` for that backend)

aluzzardi replied:
On a side note:

This is related to `swarm.discovery` and `swarm.heartbeat`: Does it make sense for cluster to only take a `Discovery`, not actually create it?

Those two options are only useful for discovery not for the cluster

vieux replied:
I agree, the thing is,  you need `swarm.heartbeat` to create the discovery, so it feels weird to parse a `swarm.` opt outside of swarm, to create the discovery and give it to swarm.

aluzzardi commented on cli/help.go:
```diff
@@ -25,7 +25,8 @@ ARGUMENTS:
 OPTIONS:
    {{range .Flags}}{{.}}
-   {{end}}{{ end }}
+   {{end}}{{if (eq .Name "manage")}}{{printf "\t * swarm.overcommit=0.05\tovercommit to apply on resources"}}
```
We don't have an option currently, but it's a bit bad to have the defaults duplicated. Out of scope with this PR.

vieux replied:
I agree

aluzzardi commented:
Minor nits and comments.

Overall LGTM - some questions regarding the relationship between Discovery and Cluster but those should probably be addressed by other PRs.

vieux commented:
@aluzzardi I know it doesn't fix the core issue, but I renamed `swarm.heartbeat` to `swarm.discovery.heartbeat` in the last commit, let me know what you think.

aluzzardi commented:
LGTM
